File: Lexer/Lexer.py
from AFD import AFD
from Token import Token
from Comment import Comment
from Reservada import Reservada
from MathOperator import MathOperator
from Number import Number
from ID import ID
from StringText import StringText

class Lexer:
    def __init__(self, code):
        self.tokens = []
        self.code = iter(code)
        self.line_error = 1
        self.column = 0
        self.pos = 0  # Variável para rastrear a posição atual do iterador
        self.afds = [Comment(), Reservada(), MathOperator(), Number(), ID(), StringText()]

    def skip_white_space(self):
        while True:
            try:
                current = next(self.code)
                self.column += 1
                if current == ' ' or current == '\n' or current == '\r' or current == '\t':
                    if current == '\n':
                        self.line_error += 1
                        self.column = 0
                else:
                    break
            except StopIteration:
                break

    def get_tokens(self):
        while True:
            try:
                self.skip_white_space()
                current = next(self.code)
                accepted = False

                for afd in self.afds:
                    self.pos += 1  # Atualizar a posição manualmente
                    pos = self.pos  # Salvar a posição atual
                    t = afd.evaluate(current, self.line_error, abs(self.pos - self.column))

                    if t is not None:
                        accepted = True
                        self.tokens.append(t)
                        break
                    else:
                        self.pos = pos  # Voltar para a posição original

                if not accepted:
                    raise RuntimeError(f"Error: Token not recognized: {current} at line: {self.line_error} column: {abs(self.pos - self.column)}")
            except StopIteration:
                break

        self.tokens.append(Token("EOF", "$", self.line_error, abs(self.pos - self.column)))
        return self.tokens

File: src/alvaro/Lexico/Lexer.java
package Lexico;

import java.text.CharacterIterator;
import java.text.StringCharacterIterator;
import java.util.ArrayList;
import java.util.List;

public class Lexer {
  private List<Token> tokens;
  private List<AFD> afds;
  private CharacterIterator code;
  private int lineError, coluna;

  public Lexer(String code) {
    tokens = new ArrayList<>();
    this.code = new StringCharacterIterator(code);
    this.lineError = 1;
    this.coluna = 0;
    afds = new ArrayList<>();
    afds.add(new Comment());
    afds.add(new Reservada());
    afds.add(new MathOperator());
    afds.add(new Number());
    afds.add(new ID());
    afds.add(new StringText());
  }

  // Metodo para pular espaco em branco
  public void skipWhiteSpace() {
    while (code.current() == ' ' || code.current() == '\n' || code.current() == '\r' || code.current() == '\t'
        || code.current() == CharacterIterator.DONE || (code.current() == '\r' && code.current() == '\n')) {
      if (code.current() == '\n' || (code.current() == '\r' && code.current() == '\n')) {
        this.lineError++;
        this.coluna = code.getIndex();
      }
      code.next();
    }
  }

  public List<Token> getTokens() {
    boolean accepted;

    while (code.current() != CharacterIterator.DONE) {
      this.coluna++;
      accepted = false;
      skipWhiteSpace();

      if (code.current() == CharacterIterator.DONE)
        break;

      for (AFD afd : afds) {
        int pos = code.getIndex();
        Token t = afd.evaluate(code, lineError, (Math.abs(code.getIndex() - this.coluna)));

        if (t != null) {
          accepted = true;
          tokens.add(t);
          break;
        } else {
          code.setIndex(pos);
        }
      }

      if (accepted)
        continue;

      throw new RuntimeException("Error: Token not recognized: " + code.current() + " at line: " + lineError
          + " column: " + (Math.abs(code.getIndex() - this.coluna)));
    }

    tokens.add(new Token("EOF", "$", lineError, (Math.abs(code.getIndex() - this.coluna))));
    return tokens;
  }
}